client/my-sites/email/email-forwarding/email-forwarding-item.jsx

```diff
@@ -86,8 +86,8 @@ class EmailForwardingItem extends React.Component {
 								em: <em />,
 							},
 							args: {
-								email: email,
-								forwardTo: destination,
+								email: email || '',
```

Comment:
What was the behavior change which made this necessary? I guess it's likely more a difference in the `sprintf-js` implementation vs. what's shipped with Jed (which I assumed was the same, based on [their documentation](https://github.com/messageformat/Jed#credits)).

Reply:
I assumed so too, but I suppose at some point things appear to have changed. The tests for this component seem to indicate that the `sprintf` that shipped with Jed would replace `undefined` with empty strings, whereas `sprintf-js` replaces them with the string `'undefined'`.

I'm not terribly fond of the latter (and there are a number complaints about it on their largely-ignored bugtracker), but it is a case of relying on undocumented behaviour. I could see either option being acceptable, just like it would be acceptable to abort the replacement and leave the placeholder in the string.

That's why I opted to make it explicit in this case, since the tests are relying on a particular behaviour.

Reply:
Do you think there might be other occurrences of this same issue elsewhere in the code which isn't covered by tests? I wonder if it might be worth exploring to have it be part of the abstraction of `translate`, even if just for now to avoid dealing with addressing all of the potential breakages.

Reply:
There can definitely be other instances of this in the codebase. In order to avoid the issue altogether, we'd need to audit every instance of a `translate` call with named arguments.

As for handling this in `translate`, I'm not sure there are any great solutions for this. The issue is that we only become aware of missing placeholders as we parse the translation string inside `sprintf`. We could search for `undefined` in the `sprintf` return and try to work out if it should legitimately be there based on the input string and the args, but that seems quite hacky and potentially slow. The alternative would be to write `sprintf` ourselves to correctly handle all our expected use-cases. Not impossible, but certainly challenging.

Reply:
Actually, there may be a much simpler solution to this if `sprintf` is making a distinction between an args array of `{ foo: undefined }` and an args array of `{}`. Let me quickly check that.

Reply:
> Actually, there may be a much simpler solution to this if `sprintf` is making a distinction between an args array of `{ foo: undefined }` and an args array of `{}`. Let me quickly check that.

No, sadly that isn't the case:

```
sprintf('%(arg1)s', {});
// "undefined"

sprintf('%(arg1)s', {arg1: undefined})
// "undefined"
```

We could look for arguments with a value of `undefined` going into `sprintf` and replace those, but we'd have no way of knowing which arguments would be missing altogether.

Reply:
So, to summarise, here are the options we can choose from:
1 - Leave things as they are and risk `undefined` making its way into strings sometimes. When that happens, fix responsible code by making the behaviour explicit.
2 - Pre-parse strings before sending them to `sprintf`, to identify the expected arguments and ensure that they're present.
3 - Analyse `sprintf` output, looking for `undefined`, and ensuring it's there for a legitimate reason.
4 - Implement our own `sprintf`.
5 - "Fork" `sprintf-js`, by creating a new `sprintf.js` file in `i18n-calypso` which is a modified copy of the BSD-licensed `sprintf-js`.

There may be other options I've overlooked.

Which one do you think is the way to go, @aduth? And for good measure, @jsnajdr? 🙂

Reply:
Hm.

The second option seems to be a good balance between simple and reliable, but might also come with some additional performance overhead of iterating the object to find `undefined` values.

The first option feels a bit too risky, and maybe too much to be taking on as part of this effort to port to Tannin (i.e. maybe a separate effort).

The third feels too error-prone and subject to false positives.

I think the main reason I hadn't implemented named parameters in `@tannin/sprintf` was because this doesn't appear to be a thing in the [standard GNU `printf` specifications](https://www.gnu.org/software/libc/manual/html_node/Formatted-Output.html#Formatted-Output). Otherwise I don't have strong reservations about building it in.

Reply:
In that case, if @jsnajdr agrees, I'll place this PR on hold while I add support for named arguments to `@tannin/sprintf`!

Reply:
Created https://github.com/aduth/tannin/pull/1 in the meantime. Testing to ensure that it works correctly with Calypso.

Reply:
The modified `@tannin/sprintf` appears to be compatible with Calypso's usage of named placeholders. It fails on some tests due to an unrelated bug; creating another PR for that.

Reply:
The changes to `@tannin/sprintf` are merged and awaiting a new release.